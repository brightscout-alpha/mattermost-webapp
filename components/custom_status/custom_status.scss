.StatusModal {
    width: 600px !important;
    margin: 30px auto !important;

    .modal-dialog {
        margin: auto;
    }

    .modal-content {
        width: 600px;
        padding: 28px 24px 0 24px;
    }

    .modal-body {
        overflow: visible;
        padding: 12px 12px 0px 12px;
    }

    .modal-footer {
        margin-top: 0;
        padding: 0px 12px 8px;
    }

    .GenericModal__header {
        padding: 0 0 20px;
    }

    .GenericModal__body {
        padding: 0;
    }

    .form-control {
        font-size: 14px;
        height: 4.8rem;
        padding: 0 5.2rem;
        font-weight: 600;
        border-radius: 4px;

        &:focus {
            box-shadow: inset 0 0 0 1px rgba(var(--button-bg-rgb), 1);
            border-color: rgba(var(--button-bg-rgb), 1);
        }
    }

    .StatusModal__input {
        position: relative;
    }
}

.StatusModal__emoji-container {
    position: absolute;
    margin-right: 2px;
    top: 6px;
    left: 10px;
}

.StatusModal__clear-container {
    position: absolute;
    top: 14px;
    right: 16px;
}

.StatusModal__emoji-button {
    background: transparent;
    border: none;
    border-radius: 4px;
    width: 36px;
    height: 36px;
    display: flex;
    align-items: center;
    justify-content: center;
    fill: rgba(var(--center-channel-color-rgb), 0.56);
    color: rgba(var(--center-channel-color-rgb), 0.56);

    &:hover {
        background: rgba(var(--center-channel-color-rgb), 0.08);
        fill: rgba(var(--center-channel-color-rgb), 0.72);
        color: rgba(var(--center-channel-color-rgb), 0.72);
    }

    &:active,
    &--active,
    &--active:hover {
        background: rgba(var(--button-bg-rgb), 0.08);
        fill: var(--button-bg);
        color: var(--button-bg);
    }

    svg {
        width: 20px;
        height: 20px;
    }
}

.StatusModal__footer {
    text-align: right;
    padding: 2rem 0;
    display: flex;
    align-items: center;
    justify-content: space-between;

    .btn {
        font-size: 14px;
        font-weight: 600;
        padding: 0 2rem;
        height: 4rem;
        border-radius: 4px;
    }
}

.ic-clear-search, .ic-search {
    background-repeat: no-repeat;
    background-position: center;
    cursor: pointer;
    position: absolute;
    z-index: 1;
}

.ic-clear-search {
    right: 12px;
    top: 12px;
    width: 2rem;
    height: 2rem;
}

.statusSuggestion  {
    height: 250px;
    overflow: auto;
    margin-top: 8px;
}

.statusSuggestion__content  {
    display: flex;
    flex-direction: column;
    height: 100%;
    width: 100%;
    box-shadow: none;
    border: none;
    overflow: auto;
}

.statusSuggestion__row {
    display: flex;
    font-size: inherit;
    min-height: 44px;
    align-items: center;
    padding: 0 20px;
    color: var(--center-channel-color);
    &:hover {
        background-color: rgba(var(--center-channel-color-rgb), 0.08);
    }

    &:active {
        background-color: rgba(var(--button-bg-rgb), 0.08);
        color: v(button-bg);
    }
}
.statusSuggestion__icon{
    margin-right: 10px;
}

.statusSuggestion__title{
    display: flex;
    font-weight: 600;
    height: 24px;
    margin-top: 8px;
}

.statusSuggestion__text{
    font-weight: 600;
}

.custom-status-button > span:first-child > span.icon {
    display: flex;
}

.custom-status-emoji > span:first-child {
    background-size: 20px;
}
