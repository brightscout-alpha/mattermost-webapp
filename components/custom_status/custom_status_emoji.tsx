// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react';
import {Tooltip} from 'react-bootstrap';
import {useSelector} from 'react-redux';

import {getCurrentUser} from 'mattermost-redux/selectors/entities/users';

import OverlayTrigger from 'components/overlay_trigger';
import RenderEmoji from 'components/emoji/render_emoji';

import Constants from 'utils/constants';
import {GlobalState} from 'types/store';

interface ComponentProps {
    emojiSize?: number;
    showTooltip?: boolean;
    tooltipDirection?: 'top' | 'right' | 'bottom' | 'left';
    emojiStyle?: React.CSSProperties,
}

const CustomStatusEmoji = ({tooltipDirection = 'bottom', showTooltip = false, emojiSize = 16, emojiStyle = {}}: ComponentProps) => {
    const currentUser = useSelector((state: GlobalState) => getCurrentUser(state));
    if (!(currentUser && currentUser.props && currentUser.props.customStatus)) {
        return null;
    }

    const customStatus = JSON.parse(currentUser.props.customStatus);
    const statusEmoji = (
        <RenderEmoji
            emoji={customStatus.emoji}
            size={emojiSize}
            emojiStyle={emojiStyle}
        />
    );

    let status = statusEmoji;
    if (showTooltip) {
        status = (
            <OverlayTrigger
                delayShow={Constants.OVERLAY_TIME_DELAY}
                placement={tooltipDirection}
                overlay={
                    <Tooltip id='custom-status-tooltip'>
                        <div className='custom-status'>
                            {statusEmoji}
                            <span className='custom-status-text'>
                                {customStatus.text}
                            </span>
                        </div>
                    </Tooltip>
                }
            >
                {statusEmoji}
            </OverlayTrigger>
        );
    }

    return status;
};

export default CustomStatusEmoji;
