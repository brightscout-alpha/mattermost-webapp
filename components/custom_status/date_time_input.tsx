// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React, {useEffect, useState} from 'react';
import {useSelector} from 'react-redux';
import DayPickerInput from 'react-day-picker/DayPickerInput';
import {DayModifiers, NavbarElementProps} from 'react-day-picker';

import moment from 'moment-timezone';

import MenuWrapper from 'components/widgets/menu/menu_wrapper';
import Menu from 'components/widgets/menu/menu';
import Timestamp from 'components/timestamp';
import {getCurrentLocale} from 'selectors/i18n';
import {Constants} from 'utils/constants';
import {getCurrentDateTimeForTimezone} from 'utils/timezone';
import {localizeMessage} from 'utils/utils';

const Navbar: React.FC<Partial<NavbarElementProps>> = (navbarProps: Partial<NavbarElementProps>) => {
    const {
        onPreviousClick,
        onNextClick,
        className,
    } = navbarProps;
    const styleLeft: React.CSSProperties = {
        float: 'left',
        fontSize: 18,
    };
    const styleRight: React.CSSProperties = {
        float: 'right',
        fontSize: 18,
    };

    return (
        <div className={className}>
            <button
                className='style--none'
                style={styleLeft}
                onClick={(e) => {
                    e.preventDefault();
                    if (onPreviousClick) {
                        onPreviousClick();
                    }
                }}
            >
                <i
                    className='fa fa-angle-left'
                    aria-hidden='true'
                />
            </button>
            <button
                className='style--none'
                style={styleRight}
                onClick={(e) => {
                    e.preventDefault();
                    if (onNextClick) {
                        onNextClick();
                    }
                }}
            >
                <i
                    className='fa fa-angle-right'
                    aria-hidden='true'
                />
            </button>
        </div>
    );
};

const {CUSTOM_STATUS_TIME_PICKER_INTERVALS_IN_MINUTES} = Constants;
export function getRoundedTime(value: Date) {
    const roundedTo = CUSTOM_STATUS_TIME_PICKER_INTERVALS_IN_MINUTES;
    console.log(value.toISOString(), new Date().toISOString())
    const start = moment(value);
    const diff = start.minute() % roundedTo;
    if (diff === 0) {
        return value;
    }
    const remainder = roundedTo - diff;
    return moment(start).add(remainder, 'm').seconds(0).milliseconds(0).toDate();
}

const getTimeInIntervals = (startTime: Date): Date[] => {
    const interval = CUSTOM_STATUS_TIME_PICKER_INTERVALS_IN_MINUTES;
    let time = startTime;
    const nextDay = moment(startTime).add(1, 'days').startOf('day').toDate();
    const intervals: Date[] = [];
    while (time < nextDay) {
        intervals.push(time);
        time = moment(time).add(interval, 'minutes').seconds(0).milliseconds(0).toDate();
    }

    return intervals;
};

type Props = {
    time: Date;
    handleChange: (date: Date) => void;
    timezone?: string;
}

const DateTimeInputContainer: React.FC<Props> = (props: Props) => {
    const locale = useSelector(getCurrentLocale);
    const {time, handleChange, timezone} = props;
    const [timeOptions, setTimeOptions] = useState<Date[]>([]);

    const setTimeAndOptions = () => {
        const currentTime = new Date();
        let startTime = moment(time).startOf('day').toDate();
        if (time.getDate() === currentTime.getDate()) {
            startTime = getRoundedTime(currentTime);
        }
        setTimeOptions(getTimeInIntervals(startTime));
    };

    useEffect(setTimeAndOptions, [time]);

    const handleDayChange = (day: Date, modifiers: DayModifiers) => {
        if (modifiers.today) {
            const currentTime = new Date();
            const roundedTime = getRoundedTime(currentTime);
            handleChange(roundedTime);
        } else {
            const dateWithProperTime = moment(day).startOf('day').toDate();
            handleChange(dateWithProperTime);
        }
    };

    const handleTimeChange = (e: React.MouseEvent, time: Date) => {
        e.preventDefault();
        handleChange(time);
    };

    const currentTime = new Date();
    const modifiers = {
        today: currentTime,
    };

    return (
        <div className='dateTime'>
            <div className='dateTime__date'>
                <span className='dateTime__input-title'>{'Date'}</span>
                <span className='dateTime__date-icon'>
                    <i className='icon-calendar-outline'/>
                </span>
                <DayPickerInput
                    value={time}
                    onDayChange={handleDayChange}
                    inputProps={{
                        readOnly: true,
                        className: 'dateTime__input',
                    }}
                    dayPickerProps={{
                        navbarElement: <Navbar/>,
                        fromMonth: currentTime,
                        modifiers,
                        locale: locale.toLowerCase(),
                        disabledDays: {
                            before: currentTime,
                        },
                    }}
                />
            </div>
            <div className='dateTime__time'>
                <MenuWrapper
                    className='dateTime__time-menu'
                >
                    <div>
                        <span className='dateTime__input-title'>{'Time'}</span>
                        <span className='dateTime__time-icon'>
                            <i className='icon-clock-outline'/>
                        </span>
                        <div
                            className='dateTime__input'
                        >
                            <Timestamp
                                useRelative={false}
                                useDate={false}
                                value={time}
                            />
                        </div>
                    </div>
                    <Menu
                        ariaLabel={localizeMessage('time_dropdown.choose_time', 'Choose a time')}
                        id='expiryTimeMenu'
                    >
                        <Menu.Group>
                            {Array.isArray(timeOptions) && timeOptions.map((option, index) => (
                                <Menu.ItemAction
                                    onClick={(e: React.MouseEvent) => handleTimeChange(e, option)}
                                    key={index}
                                    text={
                                        <Timestamp
                                            useRelative={false}
                                            useDate={false}
                                            value={option}
                                        />
                                    }
                                />
                            ))}
                        </Menu.Group>
                    </Menu>
                </MenuWrapper>
            </div>
        </div>
    );
};

export default DateTimeInputContainer;
