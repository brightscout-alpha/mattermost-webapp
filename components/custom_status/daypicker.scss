.DayPicker {
    &-wrapper &-Months &-Caption {
        text-align: center;
    }
}
