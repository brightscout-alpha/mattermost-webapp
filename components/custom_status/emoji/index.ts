// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect} from 'react-redux';

import {getCurrentUser} from 'mattermost-redux/selectors/entities/common';

import {getEmojiMap} from 'selectors/emojis';

import {GlobalState} from 'types/store';

import CustomStatusEmoji from './custom_status_emoji';

function mapStateToProps(state: GlobalState) {
    const currentUser = getCurrentUser(state);
    const emojiMap = getEmojiMap(state);
    return {
        currentUser,
        emojiMap,
    };
}

export default connect(mapStateToProps)(CustomStatusEmoji);
