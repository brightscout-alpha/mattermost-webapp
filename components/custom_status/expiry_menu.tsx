// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react';
import {FormattedMessage} from 'react-intl';

import MenuWrapper from 'components/widgets/menu/menu_wrapper';
import Menu from 'components/widgets/menu/menu';
import {localizeMessage} from 'utils/utils';
import {Duration} from 'mattermost-redux/types/users';

type ExpiryMenuItem = {
    text: string;
    value: string;
}

type Props = {
    expiry: string;
    handleExpiryChange: (event: React.MouseEvent<HTMLButtonElement, MouseEvent>, expiryValue: Duration) => void;
}

const {
    DONT_CLEAR,
    THIRTY_MINUTES,
    ONE_HOUR,
    FOUR_HOURS,
    TODAY,
    THIS_WEEK,
    DATE_AND_TIME,
} = Duration;

const ExpiryMenu: React.FC<Props> = (props: Props) => {
    const {expiry, handleExpiryChange} = props;
    const expiryMenuItems: { [key: string]: ExpiryMenuItem } = {
        [DONT_CLEAR]: {
            text: localizeMessage('expiry_dropdown.dont_clear', "Don't clear"),
            value: localizeMessage('expiry_dropdown.dont_clear', "Don't clear"),
        },
        [THIRTY_MINUTES]: {
            text: localizeMessage('expiry_dropdown.thirty_minutes', '30 minutes'),
            value: localizeMessage('expiry_dropdown.thirty_minutes', '30 minutes'),
        },
        [ONE_HOUR]: {
            text: localizeMessage('expiry_dropdown.one_hour', '1 hour'),
            value: localizeMessage('expiry_dropdown.one_hour', '1 hour'),
        },
        [FOUR_HOURS]: {
            text: localizeMessage('expiry_dropdown.four_hours', '4 hours'),
            value: localizeMessage('expiry_dropdown.four_hours', '4 hours'),
        },
        [TODAY]: {
            text: localizeMessage('expiry_dropdown.today', 'Today'),
            value: localizeMessage('expiry_dropdown.today', 'Today'),
        },
        [THIS_WEEK]: {
            text: localizeMessage('expiry_dropdown.this_week', 'This week'),
            value: localizeMessage('expiry_dropdown.this_week', 'This week'),
        },
        [DATE_AND_TIME]: {
            text: localizeMessage('expiry_dropdown.choose_date_and_time', 'Choose date and time'),
            value: localizeMessage('expiry_dropdown.date_and_time', 'Date and Time'),
        },
    };

    return (
        <div className='statusExpiry'>
            <div className='statusExpiry__content'>
                <MenuWrapper
                    className={'statusExpiry__menu'}
                >
                    <span className='expiry-wrapper expiry-selector'>
                        <FormattedMessage
                            id='expiry_dropdown.clear_after'
                            defaultMessage='Clear after'
                        />{': '}
                        <span className='expiry-value'>
                            {expiryMenuItems[expiry].value}
                        </span>
                        <span>
                            <i
                                className='fa fa-angle-down'
                                aria-hidden='true'
                            />
                        </span>
                    </span>
                    <Menu
                        ariaLabel={localizeMessage('expiry_dropdown.clear_after', 'Clear after')}
                        id='statusExpiryMenu'
                    >
                        <Menu.Group>
                            {Object.keys(expiryMenuItems).map((item) => (
                                <Menu.ItemAction
                                    key={item}
                                    onClick={(event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => handleExpiryChange(event, item)}
                                    ariaLabel={expiryMenuItems[item].text.toLowerCase()}
                                    text={expiryMenuItems[item].text}
                                    id={`expiry_menu_${item}`}
                                />
                            ))}
                        </Menu.Group>
                    </Menu>
                </MenuWrapper>
            </div>
        </div>
    );
};

export default ExpiryMenu;
