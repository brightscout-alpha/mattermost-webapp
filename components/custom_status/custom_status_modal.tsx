// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React, {useEffect, useRef, useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';
import classNames from 'classnames';
import {FormattedMessage} from 'react-intl';
import {Tooltip} from 'react-bootstrap';
import {setCustomStatus, unsetCustomStatus, removeRecentCustomStatus} from 'mattermost-redux/actions/users';
import {setCustomStatusInitialisationState} from 'mattermost-redux/actions/preferences';
import {Preferences} from 'mattermost-redux/constants';

import {UserCustomStatus} from 'mattermost-redux/types/users';

import GenericModal from 'components/generic_modal';
import EmojiIcon from 'components/widgets/icons/emoji_icon';
import EmojiPickerOverlay from 'components/emoji_picker/emoji_picker_overlay.jsx';
import {GlobalState} from 'types/store';
import OverlayTrigger from 'components/overlay_trigger';
import Constants from 'utils/constants';
import RenderEmoji from 'components/emoji/render_emoji';
import {getCustomStatus, getRecentCustomStatuses, showStatusDropdownPulsatingDot} from 'selectors/views/custom_status';

import CustomStatusSuggestion from './custom_status_suggestion';

import 'components/category_modal.scss';
import './custom_status.scss';

type Props = {
    onHide: () => void;
};

const EMOJI_PICKER_WIDTH_OFFSET = 308;
const defaultCustomStatusSuggestions: UserCustomStatus[] = [
    {emoji: 'calendar', text: 'In a meeting'},
    {emoji: 'hamburger', text: 'Out for lunch'},
    {emoji: 'sneezing_face', text: 'Out Sick'},
    {emoji: 'house', text: 'Working from home'},
    {emoji: 'palm_tree', text: 'On a vacation'},
];

const CustomStatusModal: React.FC<Props> = (props: Props) => {
    const dispatch = useDispatch();
    const currentCustomStatus = useSelector((state: GlobalState) => getCustomStatus(state));
    const recentCustomStatuses = useSelector((state: GlobalState) => getRecentCustomStatuses(state));
    const customStatusControlRef = useRef(null);
    const [showEmojiPicker, setShowEmojiPicker] = useState<boolean>(false);
    const [text, setText] = useState<string>(currentCustomStatus.text);
    const [emoji, setEmoji] = useState<string>(currentCustomStatus.emoji);
    const isStatusSet = emoji || text;
    const firstTimeModalOpened = useSelector((state: GlobalState) => showStatusDropdownPulsatingDot(state));

    const handleCustomStatusInitializationState = () => {
        if (firstTimeModalOpened) {
            dispatch(setCustomStatusInitialisationState(Preferences.CUSTOM_STATUS_MODAL_VIEWED));
        }
    };

    useEffect(handleCustomStatusInitializationState, []);

    const handleSetStatus = () => {
        const customStatus = {
            emoji: emoji || 'speech_balloon',
            text: text.trim(),
        };
        dispatch(setCustomStatus(customStatus));
    };

    const handleClearStatus = () => {
        if (currentCustomStatus.text || currentCustomStatus.emoji) {
            dispatch(unsetCustomStatus());
        }
    };

    const getCustomStatusControlRef = () => {
        return customStatusControlRef.current;
    };

    const handleEmojiClose = () => {
        setShowEmojiPicker(false);
    };

    const handleEmojiClick = (selectedEmoji: any) => {
        setShowEmojiPicker(false);
        const emojiName = selectedEmoji.name || selectedEmoji.aliases[0];
        setEmoji(emojiName);
    };

    const toggleEmojiPicker = () => {
        setShowEmojiPicker(!showEmojiPicker);
    };

    const handleTextChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        const inputText = event.target.value;
        if (inputText.length <= Constants.CUSTOM_STATUS_TEXT_CHARACTER_LIMIT) {
            setText(inputText);
        }
    };

    const handleRecentCustomStatusClear = (status: any) => {
        dispatch(removeRecentCustomStatus(status));
    };

    const customStatusEmoji = emoji || text ?
        (
            <RenderEmoji
                emoji={emoji || 'speech_balloon'}
                size={20}
            />
        ) : (
            <EmojiIcon className={'icon icon--emoji'}/>
        );

    const clearHandle = (e: React.MouseEvent<HTMLButtonElement>) => {
        e.preventDefault();
        setText('');
        setEmoji('');
    };

    const clearButton = isStatusSet ?
        (
            <div
                className='StatusModal__clear-container'
            >
                <OverlayTrigger
                    delayShow={Constants.OVERLAY_TIME_DELAY}
                    placement='top'
                    overlay={
                        <Tooltip id='clear-custom-status'>
                            {'Clear'}
                        </Tooltip>
                    }
                >
                    <button
                        className='style--none input-clear-x'
                        onClick={clearHandle}
                    >
                        <i className='icon icon-close-circle'/>
                    </button>
                </OverlayTrigger>
            </div>
        ) : null;

    const disableSetStatus = (currentCustomStatus.text === text && currentCustomStatus.emoji === emoji) ||
        (text === '' && emoji === '');

    const handleSuggestionClick = (status: any) => {
        setEmoji(status.emoji);
        setText(status.text);
    };

    const calculateRightOffSet = () => {
        let rightOffset = Constants.DEFAULT_EMOJI_PICKER_RIGHT_OFFSET;
        const target = getCustomStatusControlRef();
        if (target) {
            const anyTarget: any = target;
            rightOffset = window.innerWidth - anyTarget.getBoundingClientRect().left - EMOJI_PICKER_WIDTH_OFFSET;
            if (rightOffset < 0) {
                rightOffset = Constants.DEFAULT_EMOJI_PICKER_RIGHT_OFFSET;
            }
        }

        return rightOffset;
    };

    const recentStatuses = (
        <div>
            <div className='statusSuggestion__title'>
                {'RECENT'}
            </div>
            {
                recentCustomStatuses.map((status: UserCustomStatus) => (
                    <CustomStatusSuggestion
                        key={status.text}
                        handleSuggestionClick={handleSuggestionClick}
                        handleClear={handleRecentCustomStatusClear}
                        emoji={status.emoji}
                        text={status.text}
                    />
                ))
            }
        </div>
    );

    const renderCustomStatusSuggestions = () => {
        const recentCustomStatusTexts = recentCustomStatuses.map((status: UserCustomStatus) => status.text);
        const customStatusSuggestions = defaultCustomStatusSuggestions.
            filter((status: UserCustomStatus) => !recentCustomStatusTexts.includes(status.text)).
            map((status: UserCustomStatus, index: number) => (
                <CustomStatusSuggestion
                    key={index}
                    handleSuggestionClick={handleSuggestionClick}
                    emoji={status.emoji}
                    text={status.text}
                />
            ));

        if (customStatusSuggestions.length > 0) {
            return (
                <>
                    <div className='statusSuggestion__title'>
                        {'SUGGESTIONS'}
                    </div>
                    {customStatusSuggestions}
                </>
            );
        }

        return null;
    };

    const suggestion = (
        <div className='statusSuggestion'>
            <div className='statusSuggestion__content'>
                {recentCustomStatuses.length > 0 && recentStatuses}
                <div>
                    {renderCustomStatusSuggestions()}
                </div>
            </div>
        </div>
    );

    return (
        <GenericModal
            enforceFocus={false}
            onHide={props.onHide}
            modalHeaderText={
                <FormattedMessage
                    id='custom_status_modal_header'
                    defaultMessage='Set a status'
                />
            }
            confirmButtonText={
                <FormattedMessage
                    id='custom_status_modal_confirm'
                    defaultMessage='Set Status'
                />
            }
            cancelButtonText={
                <FormattedMessage
                    id='custom_status_modal_cancel'
                    defaultMessage='Clear Status'
                />
            }
            isConfirmDisabled={disableSetStatus}
            id='custom_status_modal'
            className={'StatusModal'}
            handleConfirm={handleSetStatus}
            handleCancel={handleClearStatus}
        >
            <div className='StatusModal__body'>
                <div className='StatusModal__input'>
                    <div
                        ref={customStatusControlRef}
                        className='StatusModal__emoji-container'
                    >
                        {showEmojiPicker && (
                            <EmojiPickerOverlay
                                target={getCustomStatusControlRef}
                                show={showEmojiPicker}
                                onHide={handleEmojiClose}
                                onEmojiClose={handleEmojiClose}
                                onEmojiClick={handleEmojiClick}
                                rightOffset={calculateRightOffSet()}
                            />
                        )}
                        <button
                            type='button'
                            onClick={toggleEmojiPicker}
                            className={classNames('emoji-picker__container', 'StatusModal__emoji-button', {
                                'StatusModal__emoji-button--active': showEmojiPicker,
                            })}
                        >
                            {customStatusEmoji}
                        </button>
                    </div>
                    <input
                        className='form-control'
                        placeholder='Set a status'
                        type='text'
                        value={text}
                        onChange={handleTextChange}
                    />
                    {clearButton}
                </div>
                {!isStatusSet && suggestion}
            </div>
        </GenericModal>
    );
};

export default CustomStatusModal;
