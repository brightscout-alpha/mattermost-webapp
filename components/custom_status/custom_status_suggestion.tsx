// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React, { useState, useRef } from 'react';
import { Tooltip } from 'react-bootstrap';

import OverlayTrigger from 'components/overlay_trigger';
import Constants from 'utils/constants';
import { CustomStatus } from 'types/store/custom_status';
import RenderEmoji from 'components/emoji/render_emoji';

import './custom_status.scss';

type Props = {
    handleSuggestionClick: (status: CustomStatus) => void;
    emoji: string;
    text: string;
    handleClear?: (status: CustomStatus) => void;
};

const CustomStatusSuggestion: React.FC<Props> = (props: Props) => {
    const { handleSuggestionClick, emoji, text, handleClear } = props;
    const [show, setShow] = useState(false);
    const textRef = useRef(null);

    const showClearButton = () => {
        setShow(true);
    };

    const hideClearButton = () => {
        setShow(false);
    };

    const handleRecentCustomStatusClear = (event: React.MouseEvent<HTMLElement>) => {
        event.stopPropagation();
        if (handleClear) {
            handleClear({
                emoji,
                text,
            });
        }
    };

    const showTextTooltip = () => {
        const element = textRef.current;
        if (element && element.offsetWidth < element.scrollWidth) {
            return true;
        }
    }

    const clearButton = handleClear ?
        (
            <div
                className='suggestion-clear'
            >
                <OverlayTrigger
                    delayShow={Constants.OVERLAY_TIME_DELAY}
                    placement='top'
                    overlay={
                        <Tooltip id='clear-recent-custom-status'>
                            {'Clear'}
                        </Tooltip>
                    }
                >
                    <span
                        className='input-clear-x'
                        onClick={handleRecentCustomStatusClear}
                    >
                        <i className='icon icon-close-circle' />
                    </span>
                </OverlayTrigger>
            </div>
        ) : null;

    let customStatusText = (
        <span className='statusSuggestion__text' ref={textRef}>
            {text}
        </span>
    );

    if (showTextTooltip()) {
        customStatusText = (
            <OverlayTrigger
                delayShow={Constants.OVERLAY_TIME_DELAY}
                placement='top'
                overlay={
                    <Tooltip id='custom-status-text'>
                        {text}
                    </Tooltip>
                }
            >
                {customStatusText}
            </OverlayTrigger>
        )
    }


    return (
        <div
            className='statusSuggestion__row cursor--pointer a11y--active'
            onMouseEnter={showClearButton}
            onMouseLeave={hideClearButton}
            onClick={
                () => handleSuggestionClick(
                    {
                        emoji,
                        text,
                    })
            }
        >
            <div className='statusSuggestion__icon'>
                <RenderEmoji
                    emoji={emoji}
                    size={20}
                />
            </div>
            {customStatusText}
            {show && clearButton}
        </div>
    );
};

export default CustomStatusSuggestion;
