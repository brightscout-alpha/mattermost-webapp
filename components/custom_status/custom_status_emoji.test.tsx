// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {mount, shallow} from 'enzyme';
import React from 'react';

import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';

import * as CustomStatusSelectors from 'selectors/views/custom_status';

import CustomStatusEmoji from './custom_status_emoji';

jest.mock('selectors/views/custom_status');

describe('components/custom_status/custom_status_emoji', () => {
    const mockStore = configureStore();
    const store = mockStore({});

    it('should match snapshot', () => {
        const wrapper = shallow(
            <Provider store={store}>
                <CustomStatusEmoji/>
            </Provider>,
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should match snapshot with props', () => {
        const wrapper = shallow(
            <Provider store={store}>
                <CustomStatusEmoji
                    emojiSize={34}
                    showTooltip={true}
                    tooltipDirection='bottom'
                />
            </Provider>,
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should not render when EnableCustomStatus in config is false', () => {
        (CustomStatusSelectors.isCustomStatusEnabled as jest.Mock).mockReturnValue(false);
        const wrapper = mount(
            <Provider store={store}>
                <CustomStatusEmoji/>
            </Provider>,
        );

        expect(wrapper.isEmptyRender()).toBeTruthy();
    });

    it('should not render when getCustomStatus returns null', () => {
        (CustomStatusSelectors.isCustomStatusEnabled as jest.Mock).mockReturnValue(true);
        (CustomStatusSelectors.getCustomStatus as jest.Mock).mockReturnValue(null);
        const wrapper = mount(
            <Provider store={store}>
                <CustomStatusEmoji/>
            </Provider>,
        );

        expect(wrapper.isEmptyRender()).toBeTruthy();
    });
});
