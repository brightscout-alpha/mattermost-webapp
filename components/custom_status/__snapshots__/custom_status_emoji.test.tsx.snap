// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/custom_status/custom_status_emoji should match snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <CustomStatusEmoji
    emojiSize={16}
    emojiStyle={
      Object {
        "marginLeft": 4,
      }
    }
    showTooltip={false}
    tooltipDirection="top"
    userID=""
  />
</ContextProvider>
`;

exports[`components/custom_status/custom_status_emoji should match snapshot with props 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <CustomStatusEmoji
    emojiSize={34}
    emojiStyle={
      Object {
        "marginLeft": 4,
      }
    }
    showTooltip={true}
    tooltipDirection="bottom"
    userID=""
  />
</ContextProvider>
`;
